pkg/agent/agent.go

```diff
@@ -29,21 +34,231 @@ import (
+			return &cni.CNIReply{Result: CniIncompatible}, err
+		}
+
+	}
+
+	// Init Port Pipeline
+	s.Pipelines[portName] = &NimbessPipeline{Mu: s.Mu, Driver: s.Driver, MetaKey: metaKey,
+		Name: fmt.Sprintf("%s_pipeline", portName), Modules: make([]network.PipelineModule, 0),
+		EgressPorts: make([]*network.EgressPort, 0)}
+	if err := s.Pipelines[portName].Init(portName, L2DriverMode, s.MetaPipelines[metaKey]); err != nil {
+		log.Error("Error while initializing Port Pipeline: %v", err)
+		return &cni.CNIReply{Result: CniIncompatible}, err
+	}
+
+	// Create Port in Nimbess Pipeline
+	eModule, err := s.Pipelines[portName].AddPort(portName, port)
+	if err != nil {
+		log.Errorf("Error while adding port %v to pipeline", port)
+		return &cni.CNIReply{}, err
+	}
+
+	log.Debugf("Rendering pipeline for port: %s", portName)
+	// Render Pipeline
+	if err := s.Driver.RenderModules(s.Pipelines[portName].Modules); err != nil {
+		return &cni.CNIReply{Result: DriverFailure}, err
+	}
+
+	log.Debug("Updating EgressPorts for this pipeline")
+	// Add all current MetaPipeline Egress ports to this port pipeline
+	for _, ePort := range s.MetaPipelines[metaKey].EgressPorts {
+		if err := s.connectEgressPort(ePort, s.Pipelines[portName]); err != nil {
+			return &cni.CNIReply{}, err
+		}
+		// Add FIB entry for this port
+		if err := s.updateFIB(s.Pipelines[portName], *ePort); err != nil {
+			return &cni.CNIReply{}, err
+		}
+	}
+
+	log.Debug("Updating EgressPorts for other pipelines")
+	// Update all other Pipelines with new Egress Port, including MetaPipeline
+	if err := s.updatePipelinesEgress(eModule, portName, s.Pipelines[portName].MetaKey); err != nil {
+		return &cni.CNIReply{}, err
+	}
+
+	cniIfIP := &cni.CNIReply_Interface_IP{
+		Version: cni.CNIReply_Interface_IP_IPV4,
+		Address: ipAddr.String(),
+	}
+
+	cniIf := &cni.CNIReply_Interface{
+		Name:        portName,
+		Mac:         port.MacAddr,
+		IpAddresses: []*cni.CNIReply_Interface_IP{cniIfIP},
+	}
+
+	return &cni.CNIReply{
+		Result:     CniOk,
+		Error:      "",
+		Interfaces: []*cni.CNIReply_Interface{cniIf},
+	}, nil
+}
+
+// updateL2FIB adds a new FIB entry to a Pipeline's forwarder
+func (s *NimbessAgent) updateL2FIB(l2Forwarder *network.Switch, macAddr string, gate network.Gate) error {
+	if err := s.Driver.AddEntryL2FIB(l2Forwarder, macAddr, gate); err != nil {
+		return err
+	}
+	l2Forwarder.L2FIB[macAddr] = gate
+	log.Infof("Added L2FIB entry: %s:%d to switch: %s", macAddr, gate, l2Forwarder.Name)
+	return nil
+}
+
+// updateFIB adds a new FIB entry to a Pipeline's forwarder for a new EgressPort
+func (s *NimbessAgent) updateFIB(pipeline *NimbessPipeline, port network.EgressPort) error {
+	log.Debugf("Updating L2FIB for pipeline %s, with port %s", pipeline.Name, port.GetName())
+	// Check if L2 or L3
+	if s.Config.Driver == L2DriverMode {
+		if s.Config.MacLearn {
+			log.Info("Ignoring static FIB update because MAC learning is enabled")
```

Comment:
why can't you do a static learn? The mac learning should overwrite it eventually unless something really bad happened. Though I guess having the ability could also get you into trouble, but it seems like you would still want both features to be able to be used at the same time.

Reply:
The option is either the data plane controls itself by learning on its own, or it gives control to the agent to decide. If we need some combination down the road we can add it, but for now we don't.